File: server/rpi_client.py
import asyncio
import websockets
import json
from datetime import datetime
import sys
import traceback
from urllib.parse import urlparse

# Get the station ID from command line arguments or use a default
if len(sys.argv) > 1:
    STATION_ID = sys.argv[1]
else:
    STATION_ID = "RPI1"  # Default ID if none provided

# Get server URL - production or development
if len(sys.argv) > 2 and sys.argv[2] == "dev":
    # Use development server with local address
    SERVER_URL = "ws://0.0.0.0:5000"
    print(f"[{datetime.now()}] DEVELOPMENT MODE: Connecting to local server")
else:
    # Production URL
    SERVER_URL = "wss://xeryonremotedemostation.replit.app"
    print(f"[{datetime.now()}] PRODUCTION MODE: Connecting to deployed server")

async def connect_to_server():
    # The server expects connections to /rpi/{rpiId}
    uri = f"{SERVER_URL}/rpi/{STATION_ID}"
    print(f"[{datetime.now()}] Connecting with URI: {uri}")
    print(f"[{datetime.now()}] Station ID: {STATION_ID}")

    # Debug output
    print(f"[{datetime.now()}] WEBSOCKET DEBUG INFO:")
    print(f"[{datetime.now()}] - Server URL: {SERVER_URL}")
    print(f"[{datetime.now()}] - Station ID: {STATION_ID}")
    print(f"[{datetime.now()}] - Complete URI: {uri}")
    print(f"[{datetime.now()}] - URI components:")
    parsed = urlparse(uri)
    print(f"[{datetime.now()}]   - Scheme: {parsed.scheme}")
    print(f"[{datetime.now()}]   - Netloc: {parsed.netloc}")
    print(f"[{datetime.now()}]   - Path: {parsed.path}")

    while True:  # Reconnect loop
        try:
            print(f"[{datetime.now()}] Connecting to {uri}...")

            # Add more detailed connection debugging
            print(f"[{datetime.now()}] Connection attempt with extra debugging...")

            # Use extra debug parameters
            async with websockets.connect(
                uri,
                extra_headers={
                    'User-Agent': 'RaspberryPi-Client/1.0',
                    'X-Station-ID': STATION_ID
                },
                ping_interval=30,
                ping_timeout=10,
                close_timeout=5
            ) as websocket:
                print(f"[{datetime.now()}] Connected to server as {STATION_ID}")

                # Send registration message
                registration_message = {
                    "status": "ready", 
                    "message": "RPi device online and ready to accept commands",
                    "type": "register",
                    "rpi_id": STATION_ID
                }
                print(f"[{datetime.now()}] Sending registration message: {registration_message}")
                await websocket.send(json.dumps(registration_message))

                # Process incoming messages
                async for message in websocket:
                    try:
                        data = json.loads(message)
                        command = data.get("command", "unknown")
                        direction = data.get("direction", "none")
                        print(f"[{datetime.now()}] Received command: {command}, direction: {direction}")

                        # Process the command here (implement your hardware control)
                        # For example, if command is "move", control the actuator

                        # Send back response
                        response = {
                            "status": "success",
                            "rpi_id": STATION_ID,
                            "message": f"Command '{command}' executed with direction '{direction}'"
                        }
                        await websocket.send(json.dumps(response))
                    except json.JSONDecodeError as e:
                        print(f"[{datetime.now()}] Invalid message: {message}")
                        print(f"[{datetime.now()}] Error: {str(e)}")
        except websockets.exceptions.ConnectionClosed as e:
            print(f"[{datetime.now()}] Connection closed: {str(e)}. Reconnecting in 5 seconds...")
            await asyncio.sleep(5)
        except websockets.exceptions.WebSocketException as e:
            print(f"[{datetime.now()}] WebSocket error: {str(e)}. Reconnecting in 5 seconds...")
            traceback.print_exc()  # Print full error for debugging
            await asyncio.sleep(5)
        except Exception as e:
            print(f"[{datetime.now()}] Unexpected error: {str(e)}. Reconnecting in 5 seconds...")
            traceback.print_exc()  # Print full error for debugging
            await asyncio.sleep(5)

if __name__ == "__main__":
    print(f"[{datetime.now()}] Starting RPI WebSocket client for {STATION_ID}")
    print(f"[{datetime.now()}] To use a different ID, run: python rpi_client.py YOUR_STATION_ID")
    print(f"[{datetime.now()}] To connect to development server, run: python rpi_client.py YOUR_STATION_ID dev")
    asyncio.run(connect_to_server())

File: server/rpi-client.ts

import WebSocket from 'ws';

// Simple RPi client simulator to test WebSocket integration
export function startRPiClient(rpiId: string, serverUrl: string = 'ws://localhost:5000') {
  const ws = new WebSocket(`${serverUrl}/rpi/${rpiId}`);
  
  ws.on('open', () => {
    console.log(`RPi Client ${rpiId} connected to server`);
    
    // Send initial status message
    const initialStatus = {
      status: 'ready',
      message: 'RPi device online and ready to accept commands'
    };
    ws.send(JSON.stringify(initialStatus));
  });
  
  ws.on('message', (data) => {
    try {
      const message = JSON.parse(data.toString());
      console.log(`RPi ${rpiId} received command:`, message);
      
      // Simulate processing and respond with success
      setTimeout(() => {
        const response = {
          status: 'success',
          message: `Executed command: ${message.command || 'unknown'}`
        };
        ws.send(JSON.stringify(response));
      }, 500); // Simulate 500ms processing time
    } catch (err) {
      console.error(`Error processing message on RPi ${rpiId}:`, err);
      ws.send(JSON.stringify({
        status: 'error',
        message: 'Failed to process command'
      }));
    }
  });
  
  ws.on('close', () => {
    console.log(`RPi Client ${rpiId} disconnected`);
    // Try to reconnect after 5 seconds
    setTimeout(() => startRPiClient(rpiId, serverUrl), 5000);
  });
  
  ws.on('error', (error) => {
    console.error(`RPi WebSocket error: ${error}`);
    ws.close();
  });
  
  return ws;
}
import WebSocket from 'ws';
import { log } from './vite';

export function startRPiClient(rpiId: string, serverUrl: string) {
  let ws: WebSocket | null = null;
  let reconnectAttempts = 0;
  const maxReconnectAttempts = 10;
  const baseReconnectDelay = 1000;

  const connect = () => {
    // Close any existing connection
    if (ws) {
      ws.close();
    }

    log(`Connecting RPi ${rpiId} to ${serverUrl}...`, 'rpi-client');
    
    // Connect to the WebSocket server
    ws = new WebSocket(serverUrl);

    ws.on('open', () => {
      log(`RPi ${rpiId} connected to server`, 'rpi-client');
      reconnectAttempts = 0;
      
      // Register the RPi with the server
      ws?.send(JSON.stringify({
        type: 'register',
        rpiId: rpiId,
        status: 'online'
      }));

      // Send heartbeat every 5 seconds
      const heartbeatInterval = setInterval(() => {
        if (ws?.readyState === WebSocket.OPEN) {
          ws.send(JSON.stringify({
            type: 'heartbeat',
            rpiId: rpiId,
            timestamp: Date.now()
          }));
        } else {
          clearInterval(heartbeatInterval);
        }
      }, 5000);
    });

    ws.on('message', (data) => {
      try {
        const message = JSON.parse(data.toString());
        log(`RPi ${rpiId} received: ${JSON.stringify(message)}`, 'rpi-client');
        
        // Handle various command types
        if (message.type === 'command') {
          // Simulate command response
          setTimeout(() => {
            ws?.send(JSON.stringify({
              type: 'commandResponse',
              rpiId: rpiId,
              commandId: message.commandId,
              status: 'success',
              data: { result: `Executed ${message.command}` }
            }));
          }, 500);
        }
      } catch (error) {
        log(`Error parsing message: ${error}`, 'rpi-client');
      }
    });

    ws.on('close', () => {
      log(`RPi ${rpiId} disconnected from server`, 'rpi-client');
      // Attempt to reconnect with exponential backoff
      if (reconnectAttempts < maxReconnectAttempts) {
        const delay = baseReconnectDelay * Math.pow(2, reconnectAttempts);
        reconnectAttempts++;
        setTimeout(connect, delay);
        log(`Attempting to reconnect in ${delay}ms (attempt ${reconnectAttempts})`, 'rpi-client');
      } else {
        log(`Max reconnect attempts reached for RPi ${rpiId}`, 'rpi-client');
      }
    });

    ws.on('error', (error) => {
      log(`WebSocket error for RPi ${rpiId}: ${error.message}`, 'rpi-client');
    });
  };

  // Initial connection
  connect();

  return {
    sendData: (data: any) => {
      if (ws?.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify(data));
        return true;
      }
      return false;
    },
    close: () => {
      if (ws) {
        ws.close();
        ws = null;
      }
    }
  };
}
